Fix terminals in editor area not reloading

ssigwart opened the pull request:
This PR fixes #140429

This isn't perfect because it just reopens then without setting the original location, but at least it reloads them.

If you have any thought on how to do that, I can try to implement it.  I think people would probably appreciate them at least reloading though.

## Testing
- Make sure `terminal.integrated.persistentSessionReviveProcess` setting is enabled.
- Other a few terminal in the terminal window.
- Quit VS Code.
- Reopen VS Code, which should reopen the terminals.

meganrogge commented on src/vs/workbench/contrib/terminal/browser/terminalService.ts:
```diff
@@ -645,7 +655,25 @@ export class TerminalService implements ITerminalService {
+				seenPersistentProcessIds.push(term.terminal);
+			}
+		}
+		const otherInstances = this.instances.filter(instance => typeof instance.persistentProcessId === 'number' && instance.shouldPersist && seenPersistentProcessIds.indexOf(instance.persistentProcessId) === -1);
```
It would be simpler to get the terminals in the editor area from `terminalEditorService`

meganrogge replied:
tried and that doesn't work, but should. 

I believe it doesn't work because of this, which happens when you close the window and then the state is saved with no terminals
https://github.com/ssigwart/vscode/blob/a084f63f4d3e5477efbcd8851008e8d996a45681/src/vs/workbench/contrib/terminal/browser/terminalEditorService.ts#L92-L103

ssigwart commented:
Thank you for reviewing this and making the updates, @meganrogge.  Please let me know if you need me to do anything more on this.

meganrogge commented:
@Tyriar when I revert ae7a4ba555b640be4d76ed482f043a51418df3bb, I don't see the duplicated prompt but I do see two terminal editors (with the same content) on reload
![Screen Shot 2022-06-17 at 7 47 19 AM](https://user-images.githubusercontent.com/29464607/174321648-f7740bd6-d4fb-48eb-92b3-97cafbfa36b4.png)

Tyriar reviewed (changes requested):
@ssigwart thanks for the PR, after looking a little closer at this I think it's the wrong approach. Instead of expanding the tabs like we do for terminal groups in the panel we want to rely on the editor's existing serialization as that will already get handled for us and it will then prevent focus/tabs appearing after start up problems.

It's been a while since I looked at this but I think there's a boolean somewhere near the [serializer](https://github.com/microsoft/vscode/blob/bfa57bdad5012d29e437e97747b4128d33e0bbf6/src/vs/workbench/contrib/terminal/browser/terminalEditorSerializer.ts) that says close the tab so it doesn't restore on _quit_ (when process _revive_ would be used) but allow is on _reload_ (when process _reconnect_ would be used).

ssigwart commented:
Thanks for the tip, @Tyriar.  It took me quite a while to come up with this new fix, but I don't think I would have gotten there without your suggestion.

The issue is `_onWillShutdown` is called before `TerminalInputSerializer` serializes, so I added `shutdownPersistentProcessId` to record the ID for use in `TerminalInputSerializer`.

After that, when restoring the editor, it was using the ID from before the shutdown, which is no longer valid.  So I added `getRevivedPtyNewId` to get the new ID.

I wasn't able to test the remote terminal settings because I wasn't really sure what "remote terminals" meant.  I assumed it had something to do with https://code.visualstudio.com/docs/remote/ssh so I tried installing the extension and connect to SSH, but it failed after trying to install a VS Code Server on the server.

Tyriar commented:
@ssigwart for remote you can create a test resolver window for the remote side of things:

![image](https://user-images.githubusercontent.com/2193314/174682802-2468d65c-7c1c-4b80-8a60-2b97e369a8b8.png)

Tyriar reviewed (changes requested):
I tested it out again, revive seems to work great but reconnect appears to have broken.

Repro:

1. Create a terminal editor on left and right
2. Type something in each
3. Reload the window

Expected: What was typed in step 2 should show up (gif off `main`)

![Recording 2022-06-20 at 15 15 51](https://user-images.githubusercontent.com/2193314/174683772-7e5be163-c744-43a1-b7ac-248cab16be8a.gif)

Actual: New processes are created because reconnect fails:

![image](https://user-images.githubusercontent.com/2193314/174683671-9bf5d768-9e99-44e9-bb19-e53eea02cd5a.png)

ssigwart commented:
Thanks, @Tyriar.  I think I got that working now.

Tyriar reviewed (approved):
Thanks, it works great 👍, it should land in v1.70
